test/widgets/compose_box_test.dart

```diff
@@ -195,10 +206,163 @@ void main() {
     });
   });
 
+  group('ComposeBox typing notices', () {
+    const narrow = TopicNarrow(123, 'some topic');
+
+    // Only test with the latest behavior of the setTypingStatus API.
+    // (Legacy variations are covered by setTypingStatus's own tests.)
+    final account = eg.account(
+      user: eg.selfUser, zulipFeatureLevel: eg.futureZulipFeatureLevel);
+
+    final contentInputFinder = find.byWidgetPredicate(
+      (widget) => widget is TextField && widget.controller is ComposeContentController);
+    final topicInputFinder = find.byWidgetPredicate(
+      (widget) => widget is TextField && widget.controller is ComposeTopicController);
+
+    void checkTypingRequest(TypingOp op, SendableNarrow narrow) =>
+      checkSetTypingStatusRequests(connection, [(op, narrow)]);
+
+    Future<void> checkStartTyping(WidgetTester tester, SendableNarrow narrow) async {
+      connection.prepare(json: {});
+      await tester.enterText(contentInputFinder, 'hello world');
+      checkTypingRequest(TypingOp.start, narrow);
+    }
+
+    testWidgets('smoke TopicNarrow', (tester) async {
+      await prepareComposeBox(tester, narrow: narrow, account: account);
+
+      await checkStartTyping(tester, narrow);
+
+      connection.prepare(json: {});
+      await tester.pump(store.typingNotifier.typingStoppedWaitPeriod);
+      checkTypingRequest(TypingOp.stop, narrow);
+    });
+
+    testWidgets('smoke DmNarrow', (tester) async {
+      final narrow = DmNarrow.withUsers(
+        [eg.otherUser.userId], selfUserId: eg.selfUser.userId);
+      await prepareComposeBox(tester, narrow: narrow, account: account);
+
+      await checkStartTyping(tester, narrow);
+
+      connection.prepare(json: {});
+      await tester.pump(store.typingNotifier.typingStoppedWaitPeriod);
+      checkTypingRequest(TypingOp.stop, narrow);
+    });
+
+    testWidgets('smoke ChannelNarrow', (tester) async {
+      const narrow = ChannelNarrow(123);
+      final destinationNarrow = TopicNarrow(narrow.streamId, 'test topic');
+      await prepareComposeBox(tester, narrow: narrow, account: account);
+
+      await tester.enterText(topicInputFinder, destinationNarrow.topic);
+      // Remove an irrelevant topic request.
+      check(connection.takeRequests()).single
```

Comment:
This confused me at first too. The response was prepared in `prepareComposeBox` whenever we call it with a `ChannelNarrow`. It would be cleaner if we just extract the `connection.prepare` out of the helper.

Reply:
Hmm, I see.

It's tricky to extract that out of the helper, though, I think. That call happens in between the setup of the store and the setup of the widgets, both of which are handled by `prepareComposeBox`.

Can this `takeRequests()` and corresponding `check` be moved into `prepareComposeBox` instead? That way, they'd live in the same place as the `connection.prepare` for the same request. I think having those two sides live in the same place is the most important aspect for having this not be confusing.

Reply:
The method that makes the API request to topics does not get called [until the first update to the topic input occurs](https://github.com/zulip/zulip-flutter/blob/f86498e21f6330d0202699f1130293878fbb6e0c/lib/widgets/autocomplete.dart#L43-L45). So moving `connection.prepare` right before `enterText` should work fine.

Reply:
That seems like it would work but also make the call sites more complex.

The fact that the topic input causes these requests is beside the point for most of these tests — it's not related to what the tests are about. So if they can avoid having to deal with it, that helps keep the tests focused.